tests/test_url.py

```diff
@@ -1657,6 +1657,78 @@ def test_join_from_rfc_3986_abnormal(url, expected):
     assert base.join(url) == expected
 
 
+EMPTY_SEGMENTS = [
+    (
+        "https://web.archive.org/web/",
+        "./https://github.com/aio-libs/yarl",
+        "https://web.archive.org/web/https://github.com/aio-libs/yarl",
+    ),
+    (
+        "https://web.archive.org/web/https://github.com/",
+        "aio-libs/yarl",
+        "https://web.archive.org/web/https://github.com/aio-libs/yarl",
+    ),
+]
+
+
+@pytest.mark.parametrize("base,url,expected", EMPTY_SEGMENTS)
+def test_join_empty_segments(base, url, expected):
+    base = URL(base)
+    url = URL(url)
+    expected = URL(expected)
+    joined = base.join(url)
+    assert joined == expected
+
+
+SIMPLE_BASE = "http://a/b/c/d"
+URLLIB_URLJOIN = [
+    ("", "http://a/b/c/g?y/./x", "http://a/b/c/g?y/./x"),
+    ("", "http://a/./g", "http://a/./g"),
+    ("svn://pathtorepo/dir1", "dir2", "svn://pathtorepo/dir2"),
+    ("svn+ssh://pathtorepo/dir1", "dir2", "svn+ssh://pathtorepo/dir2"),
+    ("ws://a/b", "g", "ws://a/g"),
+    ("wss://a/b", "g", "wss://a/g"),
+    # test for issue22118 duplicate slashes
+    (SIMPLE_BASE + "/", "foo", SIMPLE_BASE + "/foo"),
+    # Non-RFC-defined tests, covering variations of base and trailing
+    # slashes
+    ("http://a/b/c/d/e/", "../../f/g/", "http://a/b/c/f/g/"),
+    ("http://a/b/c/d/e", "../../f/g/", "http://a/b/f/g/"),
+    ("http://a/b/c/d/e/", "/../../f/g/", "http://a/f/g/"),
+    ("http://a/b/c/d/e", "/../../f/g/", "http://a/f/g/"),
+    ("http://a/b/c/d/e/", "../../f/g", "http://a/b/c/f/g"),
+    ("http://a/b/", "../../f/g/", "http://a/f/g/"),
+]
+
+
+@pytest.mark.parametrize("base,url,expected", URLLIB_URLJOIN)
+def test_join_cpython_urljoin(base, url, expected):
+    # tests from cpython urljoin
+    base = URL(base)
+    url = URL(url)
+    expected = URL(expected)
+    joined = base.join(url)
+    assert joined == expected
+
+
+URLLIB_URLJOIN_FAIL = [
+    # FIXME? - host is None
+    ("http:///", "..", "http:///"),
+    # issue 23703: don't duplicate filename
+    # FIXME? path without schema is not normalized
+    ("a", "b", "b"),
+]
+
+
+@pytest.mark.xfail(strict=True)
```

Comment:
plz always specify a `reason=` in xfail and skip markers

Reply:
Also, what's the current behavior? Does it raise an exception? Use `raises=` to specify which one.

Reply:
One problem is - 
> URL("http:///").authority # host is None
https://github.com/aio-libs/yarl/blob/6b13867652ebcfad668f6e19146cce0b7ae4bc2c/yarl/_url.py#L836
> TypeError("unsupported operand type(s) for +: 'NoneType' and 'str'")

Using "" for the host may be valid here too, but this had been evaluated already and was too dangerous - c.f. https://github.com/python/cpython/pull/15642

> URL("a").join(URL("b")) # .join uses URL.build with scheme None - 
> TypeError('NoneType is illegal for "scheme", "authority", "host", "path", "query_string", and "fragment" args, use empty string instead.')

I'm open to ideas how to work those.

Reply:
Hi,

we've had 2 remaining issues - 
- abb - `URL("a").urljoin(URL("b")) == URL("b")`
- empty-string-host - URL("http:///").urljoin("..")

abb has been solved.
For empty-string-host - I'd leave this as is, yarl has this inherited from urllib. Can't be changed without breaking lot's of things, neglect able benefit, not worth the effort without an actual use case.